elyra/pipeline/component.py

```diff
@@ -0,0 +1,233 @@
+        }
+    }
+}
+
+outputs = {
+    "id": "outPort",
+    "app_data": {
+        "ui_data": {
+            "cardinality": cardinality,
+            "label": "Output Port"
+        }
+    }
+}
+
+empty_properties = {
+    "current_parameters": {"component_source": "", "runtime_image": "", "component_source_type": ""},
+    "parameters": [{"id": "component_source"}, {"id": "runtime_image"}, {"id": "component_source_type"}],
+    "uihints": {
+        "id": "nodeProperties",
+        "parameter_info": [
+            {
+                "parameter_ref": "component_source",
+                "control": "readonly",
+                "label": {
+                    "default": "Path to Component"
+                },
+                "description": {
+                    "default": "The path to the component specification file.",
+                    "placement": "on_panel"
+                },
+                "data": {
+                    "format": "string"
+                }
+            },
+            {
+                "parameter_ref": "runtime_image",
+                "control": "custom",
+                "custom_control_id": "EnumControl",
+                "label": {
+                    "default": "Runtime Image"
+                },
+                "description": {
+                    "default": "Docker image used as execution environment.",
+                    "placement": "on_panel"
+                },
+                "data": {
+                    "items": [],
+                    "required": True
+                }
+            },
+            {
+                "parameter_ref": "component_source_type",
+                "control": "readonly",
+                "label": {
+                    "default": "Component Source Type"
+                },
+                "data": {
+                    "format": "string"
+                }
+            }
+        ],
+        "group_info": [
+            {
+                "id": "nodeGroupInfo",
+                "type": "panels",
+                "group_info": [
+                    {"id": "component_source", "type": "controls", "parameter_refs": ["component_source"]},
+                    {"id": "runtime_image", "type": "controls", "parameter_refs": ["runtime_image"]},
+                    {"id": "component_source_type", "type": "controls", "parameter_refs": ["component_source_type"]}
+                ]
+            }
+        ]
+    },
+    "resources": {}
+}
+
+
+class Property(object):
+    """
+    Represents a single property for a pipeline component
+    """
+
+    name: str
+    type: str
+    value: str
+    required: bool
+
+
+class Component(object):
+    """
+    Represents a runtime-specific component
+    """
+
+    id: str
+    name: str
+    description: str
+    runtime: str
+    properties: List[Property]
+
+    def __init__(self, id: str, name: str, description: str, runtime: Optional[str] = None, properties:
+                 List[Property] = None):
```

Comment:
```suggestion
                 Optional[List[Property]] = None) -> None:
```

Not sure what our strategy is wrt typing.  There are a number of signatures (including newly added ones) that don't have typing so its not clear how strict we should be about it.  Please advise.

Reply:
Per the in-person review, we'll tackle these in general following the release when we add tooling to validate, etc.